File: python/kruX.py
"""Matrix-based implementation for Kruskal-wallis test
Tested on Python3.2 with numpy1.7 and scipy0.12
"""

import numpy 
import scipy.stats
#import IPython
import pdb
import subprocess
import os 
import time

def calculateKruskalWallisWithMatrix(geno, 
                                     mrna,
                                     pValueThre = 1e-3,
                                     numTranscript = 1000,
                                     shuffledMrna = False):
  """Compute Kruskal-Wallis test p-values for pairs of transcripts and SNPs with
  matrix operations.
  Args:
             geno: Matrix for SNPs where each column are SNPs of a particular
                   sample and each row are the values of a particular SNP 
                   cross all samples. 
             mrna: Matrix for expressiion values of transcrtips. The format 
                   is defined same as genotype matrix.
       pValueThre: Only pairs of SNP-transcript  with p-values less than the 
                   threshold are included in the calculation
    numTranscript: The number of transcripts to be calculated in one round.
                   The parameter is determined by available  memory size  
     shuffledMrna: If T, mrna is shuffled, so the result is based on 
                   perturbation data. Default value is F
  Returns: 
     Ordered transcript-SNP pairs by their p-values
  
  version 1.0 May 2, 2013 Initial code
  """

  def preCalculate(snpsID, mrnaStart, mrnaEnd):
    """This function calculates freedome for chi-square test, numbers of SNPs 
    for no-missing  transcripts, and threshold for the corresponding p-value
    Note that the number of SNPs with a particular value
    should be corrected for missing values in transcripts
    """
    
    #mark mrna missing value 
    mrnaMissing = numpy.zeros(mrna[mrnaStart:mrnaEnd].shape)
    mrnaMissing[numpy.where(numpy.isnan(mrna[mrnaStart:mrnaEnd]))] = 1

    #Calculate the number of missing transcripts for each SNP with a value from 0,1,2 
    snpsCorrect = genoRank[:,snpsID[0],:].dot(mrnaMissing.transpose())

    #The degree of freedom of chi-square test between a pair of SNP-transcript is the number 
    #of different values of the SNP in all samples with non-missing value of the transcript
    #N is the effective number of samples for each test
    #chiSquareFD is the freedom for each test
    #Broadcasting snps[snpsID] to match snpsCorrect 
    N = snps[snpsID].transpose()[:,:,numpy.newaxis]-snpsCorrect 
    chiSquareFD = numpy.apply_along_axis(numpy.sum, 0, N>0) - 1 
    N = numpy.apply_along_axis(numpy.sum, 0, N)

    #Threshold for the chi-square test between each pair given the p-value
    chi2Thre = (scipy.stats.chi2.ppf(1-pValueThre, chiSquareFD) +  
               3 * (N + 1)) * (N * (N + 1)) / 12 
    chi2Thre[numpy.where(chiSquareFD<1)] = numpy.iinfo(numpy.int32).max

    return dict(mrna=(mrnaStart, mrnaEnd), snpsCorrect=snpsCorrect,
                N=N, chiSquareFD=chiSquareFD, chi2Thre=chi2Thre)

  def calculateNoMissing(snpsID, preResult):
    """This function handles SNPs without missing value.
    """

    #Calculate ranks for selected transcripts. NA's rank is replaced by 0
    #Note that masked array is also an option
    mrnaSelected = mrna[preResult['mrna'][0]:preResult['mrna'][1]]
    expressionRank = numpy.apply_along_axis(scipy.stats.rankdata,1,
                      mrnaSelected)
    expressionRank[numpy.where(numpy.isnan(mrnaSelected))] = 0
    
    #calulate rank statistics for different genotype values(0,1,2) and then sum
    testStat = numpy.power(numpy.dot(genoRank[:,snpsID[0],:], expressionRank.transpose()),2) / \
               (snps[snpsID].transpose()[:,:,numpy.newaxis] - preResult["snpsCorrect"])
    testStat[numpy.where(numpy.isnan(testStat))] = 0
    testStat = numpy.apply_along_axis(numpy.sum, 0,testStat) 
    preResult['testStat'] = testStat

    return

  def calculateMissing(snpsID, preResult):
    """This function handles SNPs with missing value.
    Note that only one transcript is processed in each round 
    when working on SNPs with missing values.
    """
    
    #Order SNPs according to the transcript order (nan is last) 
    #advanced indexing of genoRank
    mrnaSelected = mrna[preResult['mrna'][0]:preResult['mrna'][1]]
    mrnaOrder = numpy.argsort(numpy.ravel(mrnaSelected)) 
    genoRankOrder = genoRank[:,numpy.ravel(snpsID),:][:,:,mrnaOrder]

    #Calculate correction due to missing value in SNPs 
    sumCor = geno[numpy.ravel(snpsID),:][:,mrnaOrder]
    cumSumCor = numpy.where(numpy.isnan(sumCor), 1, 0)
    cumSumCor = numpy.cumsum(cumSumCor, axis=1)
 
    #Since missing value in the transcript are represented by 
    #zeor in order, the corresponding part in culcumSumCor is replaced by zeor, too. 
    cumSumCor = cumSumCor * (-numpy.sort(numpy.where(
                numpy.isnan(mrnaSelected), 0, -1)))
    orderStat = numpy.arange(1,1+mrnaOrder.size) * \
                (-numpy.sort(numpy.where(numpy.isnan(mrnaSelected), 0, -1)))

    #calulate rank statistics for different genotype values(0,1,2) and then sum 
    testStat= numpy.power(
                 (numpy.dot(genoRankOrder, orderStat.transpose())-
                 numpy.sum(genoRankOrder * cumSumCor[numpy.newaxis,:,:], axis=2)
                 [:,:,numpy.newaxis]), 2) / \
              (snps[snpsID].transpose()[:,:,numpy.newaxis] - 
               preResult["snpsCorrect"])
    testStat[numpy.where(numpy.isnan(testStat))] = 0
    testStat = numpy.apply_along_axis(numpy.sum, 0,testStat)
    preResult['testStat'] = testStat

    return

  def postCalculate(snpsID, preResult):
    """This function caculates P-vauels for pairs with statistics more than 
    the threshold and save the result to output list
    """

    #find the index of pairs with pvalue less then threshold
    #build record array to save information of selected pairs
    #Jun 2, 2013 Bug fixed >= instead of >
    pairSelected = numpy.where(preResult['testStat'] >= preResult['chi2Thre'])
    SNPs = numpy.ravel(snpsID)[pairSelected[0]]
    genes = numpy.arange(preResult['mrna'][0], preResult['mrna'][1])[pairSelected[1]]
    fd = preResult['chiSquareFD'][pairSelected]
    chi = (preResult['testStat'][pairSelected] * 12 / 
          (preResult['N'][pairSelected] * (preResult['N'][pairSelected] + 1))
          - 3 * (preResult['N'][pairSelected] + 1))
    pValue = scipy.stats.chi2.sf(chi, fd) #use sf to get more accrate value

    #return a structured array
    result = numpy.core.records.fromarrays([SNPs, genes, fd, chi, pValue],
             names='SNP,gene,fd,chi,pvalue')
    return result

  #list for the result
  output = list()

  #seperate SNP indexs by if having missing values
  snpsWithMissing = numpy.where(numpy.any(numpy.isnan(geno), 1)) 
  snpsNoMissing = numpy.where(numpy.all(numpy.logical_not(numpy.isnan(geno)), 1))
  
  #Genotype position index matrix  with a 3 dimensional array
  genoRank = numpy.array([numpy.where(geno==i,1,0)  
             for i in numpy.unique(numpy.ma.masked_invalid(geno).compressed())]) 

  #The number of SNPs with a particular value
  snps = numpy.sum(genoRank, axis=2).transpose()
  
  #Calculation for SNPs without missing values
  if snpsNoMissing[0].size>0:
    mrnaStart = list(range(0,mrna.shape[0],numTranscript))
    mrnaEnd   = mrnaStart[1:] + [mrna.shape[0]] 
    for ids in zip(mrnaStart, mrnaEnd):
      #Process a slice of  mrna
      preResult = preCalculate(snpsNoMissing, ids[0], ids[1])
      calculateNoMissing(snpsNoMissing, preResult) 
      result = postCalculate(snpsNoMissing, preResult) 
      output.append(result) if result.size>0 else None
  
  #Calculation for SNPs with missing values
  if snpsWithMissing[0].size>0:
    for mrnaIX in range(0, mrna.shape[0]):
      #Process a single mrna
      preResult = preCalculate(snpsWithMissing, mrnaIX, mrnaIX+1)
      calculateMissing(snpsWithMissing, preResult)
      result = postCalculate(snpsWithMissing, preResult)
      output.append(result) if result.size>0 else None

  if (output):#Bug
    output = numpy.concatenate(output)
    output = output[numpy.argsort(output['pvalue'])]
  
  return(output)


if __name__ == '__main__':
  for i in range(10):
    print(str(i))

File: trunk/r/kruX.R
CalculateKruskalWallisWithMatrix <- function(geno.para, mrna.para, 
                                             p.value.threshold = 1e-3,
                                             num.transcript = 1000,
                                             shuffled.mrna = F){
  # Compute Kruskal-Wallis test p-values for pairs of transcripts and SNPs with
  # matrix operations.
  # Tested on 2.14.1
  # Args:
  #           geno.para: Matrix for SNPs where each column are SNPs of a particular
  #                      sample and each row are the values of a particular SNP 
  #		         cross all samples. The column and row names of the matrix
  #		         represent sample names and SNP ID, respectively.
  #           mrna.para: Matrix for expressiion values of transcrtips. The format 
  #                      is defined same as genotype matrix.
  #   p.value.threshold: Only pairs of SNP-transcript  with p-values less than the threshold 
  #                      are included in the calculation
  #      num.transcript: The number of transcripts to be calculated in one round.
  #			 The parameter is determined by available  memory size  
  #       shuffled.mrna: If T, mrna is shuffled, so the result is based on 
  #                      perturbation data. Default value is F
  # Returns: 
  #   Ordered transcript-SNP pairs by their p-values
  # Date
  #   Jun 4, 2012
  #   V1.0 Modify script to handle missing value in expression values
  #        and genotype values
  #   Oct 2, 2012
  #   V1.1 Fixed bug about index variable
  #        Add parameter to support shuffle mna data

  PreCalculate <-function(SNPs.id, gene.start, gene.end){
    #This function calculates freedome for chi-square test, numbers of SNPs 
    #for no-missing  transcripts, and threshold for the corresponding p-value
    #Note that the number of SNPs with a particular value
    #should be corrected for missing values in transcripts

    genes <- row.names(mrna)[gene.start:gene.end] #Genes precessed 
    mrna.missing <- matrix(data= 0, nrow=length(genes), ncol=dim(mrna)[2]) 
    mrna.missing[is.na(mrna[genes, ])] <- 1

    #Calculate the number of missing transcripts for each SNP with a 
    #value from 0,1,2
    snps.count.correct.0  <- geno.rank.0[SNPs.id, ] %*% t(mrna.missing)
    snps.count.correct.1  <- geno.rank.1[SNPs.id, ] %*% t(mrna.missing)
    snps.count.correct.2  <- geno.rank.2[SNPs.id, ] %*% t(mrna.missing)

    #The degree of freedom of chi-square test between a pair 
    # of SNP-transcript is the number of different 
    #values of the SNP in all samples with non-missing value 
    #of the transcript
    chi.square.fd <- (((snps[SNPs.id, "0"] - snps.count.correct.0) > 0)
                     + ((snps[SNPs.id, "1"] - snps.count.correct.1) > 0) 
                     + ((snps[SNPs.id, "2"] - snps.count.correct.2) > 0)
                     - 1)
    
    #Calculate the total number of effctive samples in a test between 
    #a pair of SNP-transcript
    N <- snps[SNPs.id, "0"] - snps.count.correct.0 +
         snps[SNPs.id, "1"] - snps.count.correct.1 +
         snps[SNPs.id, "2"] - snps.count.correct.2
    ks.test.threshold<-(qchisq(p.value.threshold, chi.square.fd, lower.tail=F) +
                        3 * (N + 1)) * (N * (N + 1)) / 12
    ks.test.threshold[which(chi.square.fd < 1)] <- (2^.Machine$integer.max)

    result <-list(snps.count.correct.0 = snps.count.correct.0, 
                  snps.count.correct.1 = snps.count.correct.1,
                  snps.count.correct.2 = snps.count.correct.2,
                  chi.square.fd = chi.square.fd, N = N, 
                  ks.test.threshold = ks.test.threshold, genes = genes)
    #browser()
    return(result)
  }
  
  CalculateNoMissing <- function(SNP.id, pre.result){
    #This function handles SNPs without missing value. 

    #calculate ranks for selected transcripts
    #NA's rank is replaced by 0
    expression.rank <- apply(mrna[pre.result[["genes"]],,drop=FALSE],
                             1, rank, na.last = 'keep')
    expression.rank[which(is.na(expression.rank))] <- 0

    #calulate rank statistics for different genotype values (0,1,2) 
    #Bug fixed: add drop=FALSE, April 10, 2013
    ks.test <- matrix(0, nrow=dim(geno[SNP.id,,drop=FALSE])[1], 
                      ncol=length(pre.result[["genes"]]))
    rs.0 <- (((geno.rank.0[SNP.id, ] %*% expression.rank) ^ 2) / 
            (snps[SNP.id, "0"] - pre.result[["snps.count.correct.0"]]))
    rs.0[which(is.na(rs.0))] <- 0 #remove NaN by 0/0
    rs.1 <- (((geno.rank.1[SNP.id, ] %*% expression.rank) ^ 2) /
            (snps[SNP.id, "1"] - pre.result[["snps.count.correct.1"]]))
    rs.1[which(is.na(rs.1))] <- 0
    rs.2 <- (((geno.rank.2[SNP.id, ] %*% expression.rank) ^ 2) /
            (snps[SNP.id, "2"] - pre.result[["snps.count.correct.2"]]))
    rs.2[which(is.na(rs.2))] <- 0
    ks.test <- rs.0 + rs.1 + rs.2
    pre.result[["ks.test"]] <- ks.test
    #browser()
    return(pre.result)
  }
  
  CalculateMissing <- function(SNP.id, pre.result){
    #This function handles SNPs with missing value. 
    #It process one gene in each call. 

    #Order SNPs according transcript order (NA is last) 
    #There is only one gene.
    #browser()
    gene <- pre.result[["genes"]][1]
    geno.missing.order <- geno[SNP.id, order(mrna[gene, ]), drop=F]
    geno.rank.0.order <- geno.rank.0[SNP.id, order(mrna[gene, ])]
    geno.rank.1.order <- geno.rank.1[SNP.id, order(mrna[gene, ])]
    geno.rank.2.order <- geno.rank.2[SNP.id ,order(mrna[gene, ])]

    #Calculate correction due to missing value in SNPs and transcripts
    tmp<-matrix(0,nrow=dim(geno.missing.order)[1], 
                  ncol=dim(geno.missing.order)[2])
    tmp[which(is.na(geno.missing.order), arr.ind = T)] <- 1
    cul.sum.missing <- t(apply(tmp, 1, cumsum))
    order.seq <- seq(1,dim(mrna)[2])

    #Since missing value of in the transcript are represented by 
    #zeor in order.seq. The corresponding part in cul.sum.missing
    #is replaced by zeor, too. 
    #Note that NA is in the tail of the output of order function.
    if (T %in% is.na(mrna[gene, ])){
      cul.sum.missing[, (sum(!(is.na(mrna[gene, ]))) + 1):
                        dim(cul.sum.missing)[2]] <- 0
      order.seq[(sum(!(is.na(mrna[gene, ]))) + 1):length(order.seq)] <- 0
    } 

    #calulate rank statistics for different genotype values (0,1,2) 
    #with correction for missing value in SNPs and transcripts.
    rs.0 <- (geno.rank.0.order %*% order.seq - 
             rowSums(geno.rank.0.order * cul.sum.missing))^2 / 
             (snps[SNP.id, "0"] - pre.result[["snps.count.correct.0"]])
    rs.0[which(is.na(rs.0))]<-0 #remove NaN by 0/0
    rs.1 <- (geno.rank.1.order %*% order.seq - 
             rowSums(geno.rank.1.order * cul.sum.missing))^2 /
             (snps[SNP.id, "1"] - pre.result[["snps.count.correct.1"]])
    rs.1[which(is.na(rs.1))] <- 0
    rs.2 <- (geno.rank.2.order %*% order.seq -
             rowSums(geno.rank.2.order * cul.sum.missing))^2 /
             (snps[SNP.id, "2"] - pre.result[["snps.count.correct.2"]])
    rs.2[which(is.na(rs.2))] <- 0
    ks.test <- rs.0 + rs.1 + rs.2
    pre.result[["ks.test"]] <- ks.test

    #browser()
    return(pre.result)
  }

  PostCalculate <- function(SNP.id, pre.result){
    #This function caculates P-vauels for pairs with statistics more than 
    #the threshold and save the result to output list
    snps.selected <- which(pre.result[["ks.test"]] >= 
                           pre.result[["ks.test.threshold"]], 
                           arr.ind = T)

    #Fix bug to add drop=F April 10, 2013
    output.round <- data.frame(snps = row.names(geno[SNP.id,,drop=F])[snps.selected[, 1]],
                               gene = pre.result[["genes"]][snps.selected[, 2]],
                               fd = pre.result[["chi.square.fd"]][snps.selected],
                               chi = pre.result[["ks.test"]][snps.selected] * 
                                     12 / (pre.result[["N"]][snps.selected] * 
                                     (pre.result[["N"]][snps.selected] + 1)) - 
                                     3 * (pre.result[["N"]][snps.selected] + 1),
                               stringsAsFactors = FALSE)
    output.round$pvalue <- pchisq(output.round$chi, output.round$fd, 
                                  lower.tail=F)
    #browser()
    return(output.round)
  }
  
  #Start main program
  #Output list
  t.start = proc.time()
  output <- list()  

  #Make sure expression matrix and SNP matrix with valuse for same samples 
  geno <- geno.para[,sort(intersect(colnames(geno.para), colnames(mrna.para)))]
  mrna <- mrna.para[,sort(intersect(colnames(geno.para), colnames(mrna.para)))]

  #shuffle mrna if shuffled.mrna for FDR calculation
  if (shuffled.mrna == T){
    cat("Expression data is shuffled for FDR calculation\n")
    mrna <- mrna[, sample(seq(1, dim(mrna)[2]))]
    colnames(mrna) <- colnames(geno)
  }

  #seperate SNPs with missing values from the others
  id.missing <- apply(geno, 1, function(x){return(sum(is.na(x)))})


  #Genotype index matrix
  geno.rank.0 <- matrix(0, nrow=dim(geno)[1], ncol=dim(geno)[2])
  geno.rank.0[which(geno==0)] <- 1
  geno.rank.1 <- matrix(0, nrow=dim(geno)[1], ncol=dim(geno)[2])
  geno.rank.1[which(geno==1)] <- 1
  geno.rank.2 <- matrix(0, nrow=dim(geno)[1], ncol=dim(geno)[2])
  geno.rank.2[which(geno==2)] <- 1

  #The number of SNPs with a particular value
  snps <- matrix(data=NA, nrow=dim(geno)[1], ncol=3, 
                        dimnames=list(row.names(geno),c("0","1","2")))
  snps[, "0"] <- rowSums(geno == 0, na.rm = TRUE)
  snps[, "1"] <- rowSums(geno == 1, na.rm = TRUE)
  snps[, "2"] <- rowSums(geno == 2, na.rm = TRUE)

  index <- 0
  #Start to process SNPs without missing value
  if ( TRUE %in% (id.missing == 0)){
    t.1<-proc.time()
    id <- 0 #fix bug at version 1.1
    #In each round, only num.transcript of transcripts are processed. 
    while((id * num.transcript + 1) <= dim(mrna)[1]){
     ind.start <- id * num.transcript + 1
     ind.end <- min((id + 1) * num.transcript, dim(mrna)[1])
     cat("Processing from transcipt ID from ", ind.start, " to ", 
         ind.end," ",format(Sys.time(), "%a %b %d %X %Y"),"\n")
     pre.result <- PreCalculate(id.missing == 0, ind.start, ind.end) 
     pre.result <- CalculateNoMissing(id.missing == 0, pre.result)
     result <- PostCalculate(id.missing == 0, pre.result)
     #Chech if result is empty before adding it to output
     if (dim(result)[1] > 0){
      output[[index + 1]] <- result
      index <- index + 1
     }
     id <- id +1
    }
    t.2<-proc.time()
    #cat("Total running time for SNPS without missing values ", t.2-t.1,"\n")
  }

  #Start to process SNPs with missing value
  #Have process SNPs with missing value against each transcript
  if ( TRUE %in% (id.missing > 0)){
    t.1<-proc.time()
    for (i in seq(1, dim(mrna)[1])){
      #cat("Processing from transcipt ID ", 
      #    i ," ",format(Sys.time(), "%a %b %d %X %Y"),"\n")
      pre.result <- PreCalculate(id.missing > 0, i, i) 
      pre.result <- CalculateMissing(id.missing > 0, pre.result)
      result <- PostCalculate(id.missing > 0, pre.result)
      #Chech if result is empty before adding it to output
      if (dim(result)[1] > 0){
        output[[index + 1]] <- result
        index <- index + 1
      }
    }
    t.2<-proc.time()
    #cat("Total running time for SNPS with missing values ", t.2-t.1,"\n")
  }
  
   #cat("Total running time for R matrix-based test: ", proc.time()-t.start,"\n") 
  return(output)
}  


brutal.force.kruskal <-function(geno.matrix, mrna.matrix, p.value.threshold=1e-3 ){
  #This function calls Kruskal-Wallis. Use it to test matrix-based function
  #Parameters
  # 1. SNPs matrix
  # 2. mrna matrix
  #output
  # 1. data frame for ks test values
  t.start = proc.time()
  test.value <- data.frame()
  for (snp in row.names(geno.matrix)){
    for (gene in row.names(mrna.matrix)){
      missing <- is.na(geno.matrix[snp, ]) | is.na(mrna.matrix[gene, ])
      if (length(table(t(geno.matrix[snp, !missing]))) < 2){
        next();
      }
      k.test <- kruskal.test(t(mrna.matrix[gene, !missing]), 
                           t(geno.matrix[snp, !missing]))
      if (k.test$p.value < p.value.threshold){
      	test.value <-rbind(test.value,
                       data.frame(snps = snp, gene = gene,
                       fd = k.test$parameter, chi = k.test$statistic, 
                       pvalue = k.test$p.value, stringsAsFactors=F))
     }
    }
  }
  test.value <- test.value[order(test.value$snps, test.value$gene), ]
  cat("Total running time for R built-in test: ", proc.time()-t.start,"\n") 
  return(test.value)
}
